File: galleries/12-csv-analysis/src/main.js
// 이 예제의 핵심은 분석 결과가 아니라, 나가는 요청을 세는 일이 어디까지 믿을 만한지 보이는 데 있다.
import { getPyodide, renderPythonError, showLoading } from '../../_shared/pyodide.js';
import { addMetricRow, formatBytes } from '../../_shared/metrics.js';
import { ensureSupport } from '../../_shared/support.js';

const status = document.querySelector('#status');
const fileInput = document.querySelector('#file');
const encodingSelect = document.querySelector('#encoding');
const sampleButton = document.querySelector('#sample');
const leakButton = document.querySelector('#leak');
const leakResult = document.querySelector('#leak-result');
const netRows = document.querySelector('#net-rows');
const result = document.querySelector('#result');

const UPLOAD_PATH = '/data/uploaded.csv';

let pyodide;
let pyGlobals;
// 부팅이 끝난 시점의 요청 수와 바이트를 기억해 둔다. 이 뒤에 늘어나는 것이 곧 "나간 것" 이다.
let baseline = 0;
let baselineBytes = 0;
// 파일을 한 번이라도 올렸는지. 인코딩만 바꿔도 다시 읽으려면 알아야 한다.
let uploadedName = null;

if (ensureSupport()) {
  start();
}

async function start() {
  // 기본 버퍼는 250건이라 조용히 넘친다. 세는 것이 일인 예제이니 넉넉히 잡는다.
  performance.setResourceTimingBufferSize(1000);

  const done = showLoading(status, 'pandas 를 받는 중입니다');
  try {
    pyodide = await getPyodide({ packages: ['pandas'] });
    const response = await fetch('src/main.py');
    if (!response.ok) throw new Error(`src/main.py 를 받지 못했습니다 (${response.status})`);
    // 이 프록시는 페이지가 살아 있는 동안 계속 쓰므로 일부러 붙잡아 둔다.
    pyGlobals = pyodide.runPython(`${await response.text()}\n\nglobals()`);
    // 올린 파일이 들어갈 자리. MEMFS 라 새로고침하면 사라진다.
    pyodide.FS.mkdirTree('/data');
    done();
    const entries = requests();
    baseline = entries.length;
    baselineBytes = entries.reduce((sum, entry) => sum + entry.transferSize, 0);
    status.textContent = '준비됐습니다. CSV 를 고르거나 예제 데이터로 해 보세요.';
    for (const control of [fileInput, encodingSelect, sampleButton, leakButton])
      control.disabled = false;
    showNetwork();
  } catch (error) {
    done();
    status.replaceChildren();
    renderPythonError(status, error);
  }
}

// 이 문서가 만든 요청 전부. 브라우저가 기록해 둔 것을 그대로 읽는다.
function requests() {
  return performance.getEntriesByType('resource');
}

function showNetwork() {
  const all = requests();
  netRows.replaceChildren();
  addMetricRow(netRows, '부팅에 쓴 요청', `${baseline}건 · ${formatBytes(baselineBytes)} 받음`);

  // 버퍼는 아무나 비울 수 있는 전역 상태다. 줄었다면 세던 근거가 사라진 것이다.
  if (all.length < baseline) {
    addMetricRow(netRows, '분석하며 생긴 요청', '셀 수 없음');
    mark('계측을 믿을 수 없습니다. 기록이 지워졌습니다', false);
    return;
  }

  const since = all.slice(baseline);
  addMetricRow(netRows, '분석하며 생긴 요청', `${since.length}건`);
  if (!since.length) {
    mark('이 계측에 잡힌 요청은 없습니다', true);
    return;
  }
  const names = since.slice(0, 3).map((entry) => {
    const url = new URL(entry.name);
    // 어느 호스트로 갔는지가 유출 판정에서 제일 중요하다. 경로만 찍으면 그걸 버린다.
    return url.host === location.host ? url.pathname : `${url.host}${url.pathname}`;
  });
  const more = since.length > 3 ? ` 외 ${since.length - 3}건` : '';
  mark(`${names.join(', ')}${more} 로 나갔습니다`, false);
}

function mark(text, clean) {
  const row = addMetricRow(netRows, '관측', text);
  row.lastElementChild.className = clean ? 'verdict clean' : 'verdict';
}

function renderResult(payload) {
  const data = JSON.parse(payload);
  const parts = [];

  const heading = document.createElement('p');
  heading.textContent = `${data.title} — 행 ${data.rows.toLocaleString()}개, 열 ${data.columns.length}개`;
  parts.push(heading);

  // 열 목록은 표로 그린다. 고정폭 글꼴에 기대어 손으로 칸을 맞추면 한글에서 어긋난다.
  const table = document.createElement('table');
  table.className = 'metrics';
  const head = document.createElement('thead');
  const headRow = document.createElement('tr');
  for (const label of ['열', '타입', '고유값', '결측']) {
    const cell = document.createElement('th');
    cell.scope = 'col';
    cell.textContent = label;
    headRow.append(cell);
  }
  head.append(headRow);
  const body = document.createElement('tbody');
  for (const column of data.columns) {
    const row = document.createElement('tr');
    const name = document.createElement('th');
    name.scope = 'row';
    name.textContent = column.name;
    row.append(name);
    for (const value of [
      column.dtype,
      column.unique.toLocaleString(),
      column.missing ? `${column.missing}개` : '없음',
    ]) {
      const cell = document.createElement('td');
      cell.textContent = value;
      row.append(cell);
    }
    body.append(row);
  }
  table.append(head, body);
  parts.push(table);

  for (const [label, text] of [
    ['숫자 열 요약', data.numeric],
    [data.grouped ? `${data.grouped.column} 별로 세기` : null, data.grouped?.text],
    ['처음 세 줄', data.head],
  ]) {
    if (!label || !text) continue;
    const caption = document.createElement('p');
    caption.className = 'hint';
    caption.textContent = label;
    const block = document.createElement('pre');
    block.className = 'output';
    block.textContent = text;
    parts.push(caption, block);
  }

  result.replaceChildren(...parts);
}

async function analyze(fn, ...args) {
  const controls = [fileInput, encodingSelect, sampleButton, leakButton];
  for (const control of controls) control.disabled = true;
  result.replaceChildren(
    Object.assign(document.createElement('p'), { textContent: '읽는 중입니다…' }),
  );

  const call = pyGlobals.get(fn);
  try {
    renderResult(call(...args));
    status.textContent = '다 읽었습니다.';
  } catch (error) {
    result.replaceChildren();
    renderPythonError(result, error);
    status.textContent = '읽지 못했습니다.';
  } finally {
    call.destroy();
    for (const control of controls) control.disabled = false;
    showNetwork();
  }
}

fileInput.addEventListener('change', async () => {
  const [file] = fileInput.files;
  if (!file) return;
  status.textContent = `${file.name} (${formatBytes(file.size)}) 를 가짜 파일시스템에 올리는 중입니다…`;
  // FileReader 대신 arrayBuffer() 를 쓴다. 어차피 파일 전체를 메모리에 올린다.
  const bytes = new Uint8Array(await file.arrayBuffer());
  pyodide.FS.writeFile(UPLOAD_PATH, bytes);
  uploadedName = file.name;
  // 값을 비워 둬야 같은 파일을 다시 골랐을 때도 change 가 뜬다.
  fileInput.value = '';
  await analyze('load_uploaded', encodingSelect.value);
});

// 인코딩만 바꿔도 다시 읽어야 한다. 안 그러면 오류 화면을 띄워 놓고 고칠 길이 없다.
encodingSelect.addEventListener('change', () => {
  if (uploadedName) analyze('load_uploaded', encodingSelect.value);
});

sampleButton.addEventListener('click', () => analyze('load_sample'));

leakButton.addEventListener('click', async () => {
  const before = requests().length;
  // 응답을 읽지 않는 fetch 다. 보내는 쪽은 응답이 필요 없으니 유출 코드의 기본형이기도 하다.
  // 같은 출처의 아무 주소나 두드린다. 서버가 뭘 돌려주든 상관없다.
  fetch(`${location.pathname}?몰래=${encodeURIComponent('보낸 것처럼')}`, { method: 'POST' });
  await new Promise((resolve) => setTimeout(resolve, 1200));
  const after = requests().length;
  leakResult.textContent = [
    '응답을 읽지 않는 fetch 를 한 번 보냈습니다.',
    `Resource Timing 이 센 요청: ${before}건 → ${after}건`,
    after === before
      ? '늘지 않았습니다. 위의 표는 이 요청을 못 봅니다.'
      : '이번에는 잡혔습니다. 브라우저나 상황에 따라 다를 수 있습니다.',
    '',
    '개발자 도구의 네트워크 탭을 열어 두고 다시 눌러 보세요. 거기에는 찍힙니다.',
  ].join('\n');
  showNetwork();
});

File: galleries/12-csv-analysis/src/main.py
"""올린 CSV 를 pandas 로 훑는다. 파일은 브라우저 안의 가짜 파일시스템에서만 오간다."""

import io
import json

import pandas as pd

UPLOAD_PATH = "/data/uploaded.csv"

SAMPLE = """이름,부서,나이,연봉,입사일
김철수,개발,34,5200,2019-03-01
이영희,개발,29,4800,2021-07-15
박민수,영업,41,6100,2012-01-09
정수진,영업,36,5500,2016-11-23
최지훈,디자인,27,4200,2022-05-02
한소영,개발,45,7300,2008-04-17
"""


def load_sample() -> str:
    """파일을 안 고른 사람도 해 볼 수 있게 한다. 이것도 네트워크를 타지 않는다."""
    return describe(
        pd.read_csv(io.StringIO(SAMPLE)), "예제 데이터 (메모리에서 만든 것)"
    )


def load_uploaded(encoding: str) -> str:
    """자바스크립트가 FS 에 써 둔 파일을 읽는다. 경로만 넘겨받는다."""
    try:
        frame = pd.read_csv(UPLOAD_PATH, encoding=encoding)
    except UnicodeDecodeError as exc:
        # 한글 윈도우에서 만든 CSV 는 대개 cp949 다. 이 오류가 그 신호다.
        # from None 으로 pandas 안쪽 트레이스백을 끊는다. 안 끊으면 스무 줄이 먼저 나와서
        # 정작 무엇을 해야 하는지 적은 줄이 맨 아래로 밀린다.
        raise ValueError(
            f"{encoding} 로 읽지 못했습니다. 인코딩을 바꿔 보세요.\n원문: {exc}"
        ) from None
    return describe(frame, f"{UPLOAD_PATH} ({encoding})")


def describe(frame: pd.DataFrame, title: str) -> str:
    """화면에 뿌릴 것을 JSON 으로 돌려준다.

    열 목록을 문자열로 만들어 넘기지 않는다. 고정폭 글꼴에서 한글이 두 칸을 먹는다는
    보장이 없어서(브라우저에서 재 보면 1.44 칸이다) 손으로 맞춘 칸이 어긋난다. 표는
    자바스크립트가 <table> 로 그리게 두고 여기서는 값만 넘긴다.
    """
    columns = [
        {
            "name": str(name),
            "dtype": str(frame[name].dtype),
            "unique": int(frame[name].nunique()),
            "missing": int(frame[name].isna().sum()),
        }
        for name in frame.columns
    ]

    numeric = frame.select_dtypes("number")
    text_columns = [c for c in frame.columns if frame[c].dtype in ("object", "str")]
    grouped = None
    if text_columns and len(frame):
        # 고유값이 가장 적은 열을 고른다. 이름처럼 전부 다른 열을 세어 봐야 1 만 잔뜩 나온다.
        pick = min(text_columns, key=lambda c: frame[c].nunique())
        if frame[pick].nunique() < len(frame):
            grouped = {
                "column": str(pick),
                "text": frame[pick].value_counts().head(5).to_string(),
            }

    return json.dumps(
        {
            "title": title,
            "rows": len(frame),
            "columns": columns,
            # pandas 가 만든 표는 터미널을 겨냥한 고정폭 서식이다. 그대로 <pre> 에 넣는다.
            "numeric": numeric.describe().round(1).to_string()
            if len(numeric.columns)
            else None,
            "grouped": grouped,
            "head": frame.head(3).to_string() if len(frame) else None,
        },
        ensure_ascii=False,
    )
